src/coreComponents/mesh/CellElementRegion.cpp

```diff
@@ -35,17 +74,34 @@ CellElementRegion::CellElementRegion( string const & name, Group * const parent
 CellElementRegion::~CellElementRegion()
 {}
 
-void CellElementRegion::generateMesh( Group const & cellBlocks )
+
+void CellElementRegion::registerSubRegion( CellBlockABC const & cellBlock )
 {
   Group & elementSubRegions = this->getGroup( viewKeyStruct::elementSubRegions() );
+  // For now, subRegion name must be the same as the cell-block (so we can match them and reference them in errors).
+  CellElementSubRegion & subRegion =
+    elementSubRegions.registerGroup< CellElementSubRegion >( cellBlock.getName() );
+
+  subRegion.copyFromCellBlock( cellBlock );
+}
 
-  for( string const & cellBlockName : this->m_cellBlockNames )
+
+void CellElementRegion::generateMesh( CellElementRegionSelector & cellBlockSelector )
+{
+  std::set< string > selectedCellBlocks = cellBlockSelector.selectRegionCellBlocks( *this );
+
+  m_cellBlockNames.clear();
+  m_cellBlockNames.insert( 0, selectedCellBlocks.cbegin(), selectedCellBlocks.cend() );
+
+  for( string const & cellBlockName : m_cellBlockNames )
   {
-    CellElementSubRegion & subRegion = elementSubRegions.registerGroup< CellElementSubRegion >( cellBlockName );
-    CellBlockABC const & source = cellBlocks.getGroup< CellBlockABC >( subRegion.getName() );
-    subRegion.copyFromCellBlock( source );
+    registerSubRegion( cellBlockSelector.getCellBlock( *this, cellBlockName ) );
```

Comment:
We have a cyclic dependency here. `CellElementRegionSelector` depends on `CellElementRegion` which depends on `CellElementRegionSelector`.